# Data Understanding

The goal of this stage is to explore and understand the dataset structure, data types, target variable, and potential data quality issues before any preprocessing or modeling steps.

In [1]:
import pandas as pd
import numpy as np

pd.set_option("display.max_columns", None)

# Display formatting
from IPython.core.display import HTML
HTML("""
<style>
.dataframe table, .dataframe th, .dataframe td { font-size: 12px; }
div.output_scroll { overflow-x: auto; }
</style>
""")

In [2]:
df = pd.read_csv("telco.csv")
df.head()

,Customer ID,Gender,Age,Under 30,Senior Citizen,Married,Dependents,Number of Dependents,Country,State,City,Zip Code,Latitude,Longitude,Population,Quarter,Referred a Friend,Number of Referrals,Tenure in Months,Offer,Phone Service,Avg Monthly Long Distance Charges,Multiple Lines,Internet Service,Internet Type,Avg Monthly GB Download,Online Security,Online Backup,Device Protection Plan,Premium Tech Support,Streaming TV,Streaming Movies,Streaming Music,Unlimited Data,Contract,Paperless Billing,Payment Method,Monthly Charge,Total Charges,Total Refunds,Total Extra Data Charges,Total Long Distance Charges,Total Revenue,Satisfaction Score,Customer Status,Churn Label,Churn Score,CLTV,Churn Category,Churn Reason
0,8779-QRDMV,Male,78,No,Yes,No,No,0,United States,California,Los Angeles,90022,34.023810,-118.156582,68701,Q3,No,0,1,NaN,No,0.00,No,Yes,DSL,8,No,No,Yes,No,No,Yes,No,No,Month-to-Month,Yes,Bank Withdrawal,39.65,39.65,0.00,20,0.00,59.65,3,Churned,Yes,91,5433,Competitor,Competitor offered more data
1,7495-OOKFY,Female,74,No,Yes,Yes,Yes,1,United States,California,Los Angeles,90063,34.044271,-118.185237,55668,Q3,Yes,1,8,Offer E,Yes,48.85,Yes,Yes,Fiber Optic,17,No,Yes,No,No,No,No,No,Yes,Month-to-Month,Yes,Credit Card,80.65,633.30,0.00,0,390.80,1024.10,3,Churned,Yes,69,5302,Competitor,Competitor made better offer
2,1658-BYGOY,Male,71,No,Yes,No,Yes,3,United States,California,Los Angeles,90065,34.108833,-118.229715,47534,Q3,No,0,18,Offer D,Yes,11.33,Yes,Yes,Fiber Optic,52,No,No,No,No,Yes,Yes,Yes,Yes,Month-to-Month,Yes,Bank Withdrawal,95.45,1752.55,45.61,0,203.94,1910.88,2,Churned,Yes,81,3179,Competitor,Competitor made better offer
3,4598-XLKNJ,Female,78,No,Yes,Yes,Yes,1,United States,California,Inglewood,90303,33.936291,-118.332639,27778,Q3,Yes,1,25,Offer C,Yes,19.76,No,Yes,Fiber Optic,12,No,Yes,Yes,No,Yes,Yes,No,Yes,Month-to-Month,Yes,Bank Withdrawal,98.50,2514.50,13.43,0,494.00,2995.07,2,Churned,Yes,88,5337,Dissatisfaction,Limited range of services
4,4846-WHAFZ,Female,80,No,Yes,Yes,Yes,1,United States,California,Whittier,90602,33.972119,-118.020188,26265,Q3,Yes,1,37,Offer C,Yes,6.33,Yes,Yes,Fiber Optic,14,No,No,No,No,No,No,No,Yes,Month-to-Month,Yes,Bank Withdrawal,76.50,2868.15,0.00,0,234.21,3102.36,2,Churned,Yes,67,2793,Price,Extra data charges


In [3]:
df.shape

(7043, 50)

In [4]:
df.columns.tolist()

['Customer ID',
 'Gender',
 'Age',
 'Under 30',
 'Senior Citizen',
 'Married',
 'Dependents',
 'Number of Dependents',
 'Country',
 'State',
 'City',
 'Zip Code',
 'Latitude',
 'Longitude',
 'Population',
 'Quarter',
 'Referred a Friend',
 'Number of Referrals',
 'Tenure in Months',
 'Offer',
 'Phone Service',
 'Avg Monthly Long Distance Charges',
 'Multiple Lines',
 'Internet Service',
 'Internet Type',
 'Avg Monthly GB Download',
 'Online Security',
 'Online Backup',
 'Device Protection Plan',
 'Premium Tech Support',
 'Streaming TV',
 'Streaming Movies',
 'Streaming Music',
 'Unlimited Data',
 'Contract',
 'Paperless Billing',
 'Payment Method',
 'Monthly Charge',
 'Total Charges',
 'Total Refunds',
 'Total Extra Data Charges',
 'Total Long Distance Charges',
 'Total Revenue',
 'Satisfaction Score',
 'Customer Status',
 'Churn Label',
 'Churn Score',
 'CLTV',
 'Churn Category',
 'Churn Reason']

In [5]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7043 entries, 0 to 7042
Data columns (total 50 columns):
 #   Column                             Non-Null Count  Dtype  
---  ------                             --------------  -----  
 0   Customer ID                        7043 non-null   object 
 1   Gender                             7043 non-null   object 
 2   Age                                7043 non-null   int64  
 3   Under 30                           7043 non-null   object 
 4   Senior Citizen                     7043 non-null   object 
 5   Married                            7043 non-null   object 
 6   Dependents                         7043 non-null   object 
 7   Number of Dependents               7043 non-null   int64  
 8   Country                            7043 non-null   object 
 9   State                              7043 non-null   object 
 10  City                               7043 non-null   object 
 11  Zip Code                           7043 non-null   int64

In [6]:
df.describe().round(2).T

,count,mean,std,min,25%,50%,75%,max
Age,7043.0,46.51,16.75,19.00,32.00,46.00,60.00,80.00
Number of Dependents,7043.0,0.47,0.96,0.00,0.00,0.00,0.00,9.00
Zip Code,7043.0,93486.07,1856.77,90001.00,92101.00,93518.00,95329.00,96150.00
Latitude,7043.0,36.20,2.47,32.56,33.99,36.21,38.16,41.96
Longitude,7043.0,-119.76,2.15,-124.30,-121.79,-119.60,-117.97,-114.19
Population,7043.0,22139.60,21152.39,11.00,2344.00,17554.00,36125.00,105285.00
Number of Referrals,7043.0,1.95,3.00,0.00,0.00,0.00,3.00,11.00
Tenure in Months,7043.0,32.39,24.54,1.00,9.00,29.00,55.00,72.00
Avg Monthly Long Distance Charges,7043.0,22.96,15.45,0.00,9.21,22.89,36.39,49.99
Avg Monthly GB Download,7043.0,20.52,20.42,0.00,3.00,17.00,27.00,85.00


In [7]:
df.describe(include="object").T

,count,unique,top,freq
Customer ID,7043,7043,3186-AJIEK,1
Gender,7043,2,Male,3555
Under 30,7043,2,No,5642
Senior Citizen,7043,2,No,5901
Married,7043,2,No,3641
Dependents,7043,2,No,5416
Country,7043,1,United States,7043
State,7043,1,California,7043
City,7043,1106,Los Angeles,293
Quarter,7043,1,Q3,7043


In [8]:
df["Churn Label"].value_counts()

Churn Label
No     5174
Yes    1869
Name: count, dtype: int64

In [9]:
df["Churn Label"].value_counts(normalize=True) * 100

Churn Label
No     73.463013
Yes    26.536987
Name: proportion, dtype: float64

In [10]:
df.isna().sum().sort_values(ascending=False)

Churn Reason                         5174
Churn Category                       5174
Offer                                3877
Internet Type                        1526
Senior Citizen                          0
Married                                 0
Age                                     0
Under 30                                0
Country                                 0
State                                   0
City                                    0
Zip Code                                0
Latitude                                0
Longitude                               0
Dependents                              0
Number of Dependents                    0
Customer ID                             0
Gender                                  0
Number of Referrals                     0
Referred a Friend                       0
Quarter                                 0
Population                              0
Avg Monthly Long Distance Charges       0
Multiple Lines                    

In [11]:
df.nunique().sort_values()

State                                   1
Quarter                                 1
Country                                 1
Gender                                  2
Dependents                              2
Under 30                                2
Married                                 2
Senior Citizen                          2
Online Backup                           2
Online Security                         2
Premium Tech Support                    2
Device Protection Plan                  2
Phone Service                           2
Multiple Lines                          2
Internet Service                        2
Referred a Friend                       2
Streaming TV                            2
Streaming Movies                        2
Unlimited Data                          2
Streaming Music                         2
Churn Label                             2
Paperless Billing                       2
Internet Type                           3
Payment Method                    

## Initial Observations

- Dataset contains 7043 rows and 50 columns.
- Target variable is Churn Label.
- The dataset is moderately imbalanced.
- Some columns may cause data leakage.
- Several missing values appear to represent non-applicable cases.
- Some features have only one unique value.
- Column names contain spaces and inconsistent formatting and will be standardized during the data cleaning stage.